qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,411 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""SELECT and PREPARE for the molecular tensor hypercontraction (THC) hamiltonian"""
+from functools import cached_property
+from typing import Dict, Tuple
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft.algos.arithmetic_gates import LessThanEqualGate, LessThanGate
+from cirq_ft.algos.multi_control_multi_target_pauli import MultiControlPauli
+from cirq_ft.algos.select_swap_qrom import SelectSwapQROM
+from cirq_ft.linalg.lcu_util import preprocess_lcu_coefficients_for_reversible_sampling
+
+from qualtran import Bloq, BloqBuilder, Register, Signature, SoquetT
+from qualtran.bloqs.arithmetic import EqualsAConstant, GreaterThanConstant, ToContiguousIndex
+from qualtran.bloqs.basic_gates import Hadamard, Ry, Toffoli, XGate
+from qualtran.bloqs.on_each import OnEach
+from qualtran.bloqs.swap_network import CSwapApprox
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+def add_from_bloq_register_flat_qubits(
+    bb: 'BloqBuilder', cirq_bloq: Bloq, **regs: SoquetT
+) -> Tuple[SoquetT, ...]:
+    """Helper function to split / join registers for cirq gates expeciting single 'qubits' register.
+
+    Args:
+        bb: Bloq builder used during decompostion.
+        cirq_bloq: A CirqGateAsBloq wrapped arithmetic gate.
+        regs: bloq registers we wish to use as flat list of qubits for cirq gate.
+
+    Returns:
+        regs: bloq registers appropriately rejoined following split.
+    """
+    flat_regs = []
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            flat_regs.append([v])
+        else:
+            flat_regs.append(bb.split(v))
+    qubits = np.concatenate(flat_regs)
+    qubits = bb.add(cirq_bloq, qubits=qubits)
+    out_soqs = {}
+    start = 0
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            end = start + 1
+            out_soqs[v] = qubits[start:end][0]
+            start += 1
+        else:
+            end = start + v.reg.bitsize
+            out_soqs[v] = bb.join(qubits[start:end])
+            start += v.reg.bitsize
+    return tuple(s for _, s in out_soqs.items())
+
+
+def add_from_bloq_registers(
+    bb: 'BloqBuilder', cirq_bloq: Bloq, **bloq_regs: SoquetT
```

Comment:
is cirq_bloq supposed to be a `CirqGateAsBloq`. If yes, update annotation

Reply:
Done.